Add solax inverter model selector in config_flow to bypass discovery

brew-your-own opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a selector for the type of Inverter in the solax integration config flow.
This will bypass the inverter discovery code which tends to crash some of the dongles out there.
This was made possible by release 3.1.0 in the upstream solax library which introduced a revamped
discovery code and the ability to only discovery a subset of all inverters, see 
https://github.com/squishykid/solax/pull/145

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: #66617
- This PR is related to issue: #66617

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

brew-your-own committed:
solax: add Inverter model selector in config_flow to bypass discovery (#66617). Adapt code to upstream solax-3.1.0

brew-your-own committed:
Merge branch 'dev' into solax-inverter-selector

Darsstar reviewed (approved):
You beat me to it 😄

PS. you might want to remove the spaces before/after the `x`s in the `Checklist` section. Github isn't handling that correctly.

Darsstar reviewed (changes requested):
Sorry, I meant to click 'Request changes', not 'Approve'...

brew-your-own commented on homeassistant/components/solax/__init__.py:
```diff
@@ -1,26 +1,42 @@
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up the sensors from a ConfigEntry."""
 
+    registry_hash = {cls.__name__: cls for cls in REGISTRY}
```
ah yes, I am learning python (I am an ex-java dev) and did not realize the entry points could be used that way. You are right, then this hash is redundant. 
I am willing to try to improve it (to learn) but cannot get to it before this weekend.

brew-your-own replied:
I pushed some changes to do as you suggest

brew-your-own commented on homeassistant/components/solax/__init__.py:
```diff
@@ -1,26 +1,42 @@
+    registry_hash = {cls.__name__: cls for cls in REGISTRY}
+
+    invset = set()
+    for cls_name in entry.data[CONF_SOLAX_INVERTER]:
```
yes good point
async_migrate_entry() is a good approach - I did not know about this callback
then I guess I can do this and test it. Not before the weekend :)
If you already have some code, happy to reuse.

Darsstar commented on homeassistant/components/solax/config_flow.py:
```diff
@@ -5,37 +5,59 @@
-        data[CONF_IP_ADDRESS], data[CONF_PORT], data[CONF_PASSWORD]
+    _LOGGER.debug("CONF_SOLAX_INVERTER entry: %s", data[CONF_SOLAX_INVERTER])
+
+    invset = set()
+    for cls_name in data[CONF_SOLAX_INVERTER]:
+        invset.add(REGISTRY_HASH[cls_name])
+    inverter = await discover(
+        data[CONF_IP_ADDRESS],
+        data[CONF_PORT],
+        data[CONF_PASSWORD],
+        inverters=invset,
     )
```
I should have fixed the main cause behind the randon crashing during discovery in 3.1.0. But since [X1 Mini V34 can also handle the response from an X1 Boost](https://github.com/squishykid/solax/pull/119#issuecomment-1646691601) it would, in my opinion, be best to detect such cases and have the user deselect X1 Mini V34 in this example.

Darsstar replied:
I forgot to mention I only want to use `ALL_COMPLETED` only during initial configuration / [reconfiguration](https://developers.home-assistant.io/blog/2024/03/21/config-entry-reconfigure-step/). In `async_setup_entry()` I want to use `FIRST_COMPLETED` and pass it a single Inverter class.

brew-your-own replied:
ALL_COMPLETED does not work on my installation, I get timeouts with the following code I use to test the upstream solax library:
`    inverters = loop.run_until_complete(solax.discover("solax.home.arpa", 80, "XXXXXXXXXX", return_when=asyncio.ALL_COMPLETED))
`
I think my inverter (X1HybridG4) crashes in this case. With FIRST_COMPLETED it works.

brew-your-own committed:
improving use of entry_points
better handling of missing inverter config entry
multiple=false in config_flow for inverter selector

brew-your-own committed:
Merge branch 'home-assistant:dev' into solax-inverter-selector

brew-your-own commented:
made a few more changes following your feedback, WDYT?
I could not make ALL_COMPLETED work on my installation, I think some inverters do not like it...

brew-your-own commented on homeassistant/components/solax/config_flow.py:
```diff
@@ -5,37 +5,59 @@
 DEFAULT_PASSWORD = ""
+DEFAULT_INVERTER = ""
+
+REGISTRY_HASH = {cls.__name__: cls for cls in REGISTRY}
```
I made the change in the latest commits, see if this looks better?

brew-your-own committed:
Merge branch 'dev' into solax-inverter-selector

Darsstar reviewed (approved):
Looks good.

Mind opening an issue on the solax repo about ALL_COMPLETED not working for you?

brew-your-own committed:
Merge branch 'dev' into solax-inverter-selector

emontnemery commented:
@brew-your-own the PR has merge conflicts, can you fix them or rebase the PR please?

frenck commented:
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍

../Frenck
